### PR TITLE
Initial implementation of gentbl as a Starlark rule

### PR DESCRIPTION
**Pull request by GMNGeoffrey**
This should avoid having to list all transitive dependencies in every
invocation of gentbl and enable much nicer includes specification. This
is designed to allow a drop-in replacement to ease transition to the
new functionality. All existing usages of gentbl should also work with
this rule. I tested out fully transitioning the current build rules to
use td_library, but am going to submit this as just the change to
the rule, with the BUILD file changes coming later to make things more
incremental.

### CODE REVIEW

**Comment by chandlerc** [comment]
I would really suggest getting @hlopko to review (can't properly @ him here for some reason). He understands this part of Bazel and Starlark much better than I do.

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [inline llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: Nits: The comment is stale, and you could use doc attribute to specify documentation in a way that can be later extracted by stardoc (https://docs.bazel.build/versions/master/skylark/lib/globals.html#parameters-23)

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [inline llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl:26]: I'd vote for not allow depending on filegroups now that there is a real rule with its own provider - that's why you look into DefaultInfo, to support filegroups right? DefaultInfo is provided by so many rules and nothing would stop me from depending on something wrong like cc_binary. Wdyt?

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [inline llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl:110]: Can `srcs` be empty? If not, `allow_empty` = False`

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [inline llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: Depending on your opinion on depending on filegroup we could add a restriction that only rules providing TdFiles can go into deps. Also for gentbl_rule.

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [inline llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: `depset.to_list()` is frowned upon :) Flattening the depset can be source of slowness in larger builds. Let's not trigger djasper@ :) Use https://docs.bazel.build/versions/master/skylark/lib/Args.html and https://docs.bazel.build/versions/master/skylark/lib/Args.html#add_all.

Instead of `["-I%s" % (include,) for include in trans_includes.to_list()]`
you could write `args.add_all(trans_includes, format_each = "-I%s")`

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [inline llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: No need to run to_list there

**Review by hlopko — APPROVED** [review]
Looks really good! Just a few comments. Generally LGTM after to_list() is fixed.

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (GMNGeoffrey) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl:26]: Yeah that was the intent and I went back and forth on whether that was a good idea. When I was moving over all the rules in this file it made things a bit easier because I had an intermediate state that worked. I think it might also ease the rollout because someone won't have to switch over all their dependencies in order to adopt the new format.

What about leaving filegroups allowed here for now with a TODO to remove them once a reasonable transition period has passed?

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (GMNGeoffrey) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl:110]: A rule with only deps seems reasonable, no?

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (GMNGeoffrey) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: I think it makes sense to do that for gentbl_rule regardless (which I should probably rename, actually) since deps was not a pre-existing attribute. For this one, see my comment above.

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (GMNGeoffrey) [inline llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl:45]: @joker-eph can you comment on what the use case for this target is?

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl:26]: Your call, TODO sgtm.

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl:110]: (I have no clue what tblgen is so maybe it does, I wouldn't know :)

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: That makes good sense!

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [inline llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: I wonder if there's a more intention revealing name, wdyt about resolve_includes?

**Review by hlopko — APPROVED** [review]
Looks great, just a nit.

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (GMNGeoffrey) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: Sure. I'm also trying to find a better way to get the current package path and current repository path. `native.package_name()` and `native.repository_name()` can only be called during loading. Basically I want to be able to replicate the includes parameters from `cc_library` but AFAICT that isn't actually implemented in Starlark. Like `strip_include_prefix` treats absolute paths as relative to the repository (https://docs.bazel.build/versions/master/be/c-cpp.html#cc_library.strip_include_prefix), but I can't figure out how to do that. Is there something that has this functionality implemented in Starlark?

Comment (GMNGeoffrey) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: Ok similarly, the way I have this implemented doesn't work. I'm testing this out in IREE and the package_path I'm getting is just `llvm-project/mlir/` as opposed to `external/llvm-project/mlir/`. How do I get the actual path here? The documentation for root says that they are "are mapped together into a single directory tree to form the execution environment"
(https://docs.bazel.build/versions/master/skylark/lib/root.html), which suggests to me that I shouldn't have to be doing all this futzing with various paths, but that doesn't seem to be the case.

Comment (GMNGeoffrey) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: I came up with the solution of adding the BUILD file as a private dep in the rule

```bzl
"_build_file": attr.label(default = ":BUILD", allow_single_file = True),
```

is that the best we've got?

Comment (GMNGeoffrey) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: I found `ctx.label`. I still feel like I'm doing a silly amount of path munging here.

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [inline llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: Super nit: I'd remove the comment

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [inline llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: Only FYI: there's also ctx.bin_dir, which unless a bazel flag is used, is the same as ctx.genfiles_dir. By default ctx.actions.declare_file puts the file in the bin_dir. Sadly, internally in Google they are not the same. Just so you know in case you are wondering why some file is in genfiles in bazel but it's not in blaze.

**Review by hlopko — APPROVED** [review]

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (GMNGeoffrey) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: It's a breadcrumb for future people who are like me and would try to make this a single `add_all` with a map_each expanding to two include paths. I think it's worth saving them the effort.

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (GMNGeoffrey) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: Yes I encountered this when porting to Blaze, which is why I added `output_to_genfiles` to this rule so I get consistent behavior in both. Does that seem like the correct workaround? I think only things in genfiles can be used by later rules, right? So since this is creating header files and such to be included later I do want it to output to genfiles (although as you note, by default in Bazel there's no difference):

> output_to_genfiles
> default = False
If true, the files will be generated in the genfiles directory instead of the bin directory. Unless you need it for compatibility with existing rules (e.g. when generating header files for C++), do not set this flag.

https://docs.bazel.build/versions/master/skylark/lib/globals.html#rule.output_to_genfiles

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: Oh right I didn't realize that the intent was to have one add_all call. Then the comment makes sense you're right. Thanks for explaining.

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (hlopko) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: I'm not sure I follow "I think only things in genfiles can be used by later rules, right?". Typically everything works fine when everything uses binfiles. If some rule uses genfiles, for example it generates a C++ source file there, then other rules need to make sure they work with genfiles. In the generated C++ source file example it would mean to correctly pass `-I` or `-iquote` for both genfiles and binfiles (and non-generated source files too).

Comment (hlopko) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: So maybe at the end the problem will be solved by adding 3 includes by default - `include`, `ctx.genfiles_dir.path + "/" + include`, and `ctx.binfiles_dir.path + "/" + include`.

Comment (hlopko) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: Alternatively, we could not remember include directories in the provider, but all the possible included files? And then map files to their parent directories when creating args. Am I making sense? I am rushing this reply on a Friday afternoon :)

**Review by hlopko — APPROVED** [review]

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (GMNGeoffrey) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: Ah, that documentation made me thing that maybe if you were generating a C++ header you needed to make sure it output to genfiles. Yeah perhaps we need to add all three.

> Alternatively, we could not remember include directories in the provider, but all the possible included files? And then map files to their parent directories when creating args.

Well we already have all possible included files (that's the deps), but we still need to know the include search paths because the way in which they're included is going to be relative to those, not just by their basename. That is, we need to add the `mlir/include` directory to the include paths so that we can reference `mlir/include/mlir/IR/OpBase.td` as `mlir/IR/OpBase.td`. This is basically the same as how the C++ compiler works (which is why it would be super helpful if there was a Starlark implementation of that I could reference).

Comment (GMNGeoffrey) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl]: Ok yeah, there are a number of rules that expect generated outputs to be generated in genfiles (since that's where genrule puts them). I think it may still be good to make this rule more tolerant and have its inclusion paths prefixed with all three, but also I think it needs to output to genfiles.

File: llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl
Comment (GMNGeoffrey) [reply llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl:45]: Chatted in another forum.

**Pull request merged**
